File: python/mirage/commands/builtin/generic/cmp.py
from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass

from mirage.commands.builtin.constants import (CMP_SIZE_UNITS, INTMAX,
                                               XSTRTOUMAX_PATTERN)
from mirage.commands.builtin.utils.constants import STDIN_OPERAND
from mirage.commands.builtin.utils.output import format_records
from mirage.commands.builtin.utils.size_suffix import parse_base0
from mirage.commands.builtin.utils.stream import is_stdin, stdin_bytes
from mirage.commands.config import CommandOpts
from mirage.commands.errors import UsageError
from mirage.commands.spec import SPECS
from mirage.commands.spec.flag_view import FlagView
from mirage.commands.spec.types import CommandName, FlagValue
from mirage.commands.spec.usage import (extra_operand_error,
                                        missing_operand_error, usage_hint)
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec
from mirage.utils.errors import FS_ERRORS, format_fs_error

_TRY_HELP = "\n" + usage_hint(CommandName.CMP)
_NEWLINE = ord(b"\n")


def parse_count(raw: str, option: str, shown: str | None = None) -> int:
    """One GNU ``cmp`` byte count read the way xstrtoumax reads it.

    Base 0, so ``010`` is 8 and ``0x400`` is 1024; one leading ``+`` and
    leading whitespace are allowed; the remainder is a size suffix from
    ``cmp``'s own letter set. Every rejection -- unparsable digits,
    unknown suffix, or a product past ``INTMAX`` -- is the same usage
    error naming the long option, not a crash and not od's "too large".

    Args:
        raw (str): the count to read.
        option (str): the long option name for the diagnostic, e.g.
            ``--bytes``.
        shown (str | None): the spelling to name in the diagnostic when
            it differs from ``raw``. GNU prints the operand from the
            position it was reading, so a bad ``SKIP1`` names the whole
            ``SKIP1:SKIP2`` pair while a bad ``SKIP2`` names only itself.

    Raises:
        UsageError: the operand is not a count cmp accepts.
    """
    named = raw if shown is None else shown
    error = UsageError(f"cmp: invalid {option} value '{named}'{_TRY_HELP}")
    match = XSTRTOUMAX_PATTERN.match(raw)
    if match is None:
        raise error
    digits, suffix = match.group(1), match.group(2)
    if suffix and suffix not in CMP_SIZE_UNITS:
        raise error
    count = parse_base0(digits) * (CMP_SIZE_UNITS[suffix] if suffix else 1)
    if count > INTMAX:
        raise error
    return count


def parse_skip(raw: str) -> tuple[int, int]:
    """The ``-i`` operand as one skip per file.

    GNU takes ``SKIP`` for both files or ``SKIP1:SKIP2`` for one each,
    so ``-i 0:3`` compares all of the first file against the fourth
    byte onward of the second. A colon is the only place the first
    count may stop, which is why ``1b:1`` is rejected naming the whole
    pair while ``1:1b`` is rejected naming just ``1b``.

    Args:
        raw (str): the ``-i`` operand as typed.
    """
    first, sep, second = raw.partition(":")
    head = parse_count(first, "--ignore-initial", raw)
    if not sep:
        return head, head
    return head, parse_count(second, "--ignore-initial")


def visible(byte: int) -> str:
    """One byte rendered the way GNU ``cmp -b`` renders it.

    The cat -v alphabet: a control byte becomes ``^X`` (so tab is
    ``^I``, unlike ``cat -v`` itself), DEL becomes ``^?``, and a high
    byte becomes ``M-`` followed by the same rules on its low seven
    bits.

    Args:
        byte (int): the byte value.
    """
    if byte >= 128:
        return "M-" + visible(byte - 128)
    if byte == 127:
        return "^?"
    if byte < 32:
        return f"^{chr(byte + 64)}"
    return chr(byte)


def offset_width(sizes: list[int], limit: int | None) -> int:
    """The width GNU ``cmp -l`` pads its offset column to.

    GNU sizes the column for the largest offset it could print: the
    ``-n`` limit, cut to the bytes left in each regular file after its
    skip. A stream has no size to cut by, so a line comparing two of
    them pads to the width of the largest file offset.

    Args:
        sizes (list[int]): bytes left after the skip in each regular
            operand; a stream contributes none.
        limit (int | None): the ``-n`` count, if given.
    """
    most = min([limit if limit is not None else INTMAX, *sizes])
    return len(str(max(most, 0)))


async def cmp_cmd(
        paths: list[PathSpec],
        *,
        read_bytes: Callable[..., Awaitable[bytes]],
        stdin: ByteSource | None = None,
        silent: bool = False,
        verbose: bool = False,
        limit: int | None = None,
        print_bytes: bool = False,
        skip: tuple[int, int] = (0, 0),
) -> tuple[ByteSource | None, IOResult]:
    if len(paths) > 2:
        raise extra_operand_error(CommandName.CMP, paths[2].raw_path
                                  or paths[2].virtual)
    if not paths:
        raise missing_operand_error(CommandName.CMP, None)
    # A lone FILE1 is compared with stdin, which GNU names `-`.
    p0, p1 = paths[0], paths[1] if len(paths) > 1 else STDIN_OPERAND
    if is_stdin(p0) and is_stdin(p1):
        # Both name the one stdin: GNU sees the same file at the same
        # offset and answers equal without reading, whatever the skips.
        return None, IOResult()
    names = (p0.raw_path or p0.virtual, p1.raw_path or p1.virtual)
    read = stdin_bytes(read_bytes, stdin)
    try:
        data1 = await read(p0)
        data2 = await read(p1)
    except FS_ERRORS as exc:
        # GNU cmp reserves exit 1 for "files differ"; trouble (a missing
        # or unreadable operand) is exit 2.
        return None, IOResult(exit_code=2,
                              stderr=format_fs_error("cmp", exc, paths))
    sizes = [len(data1) - skip[0]] if not is_stdin(p0) else []
    if not is_stdin(p1):
        sizes.append(len(data2) - skip[1])
    data1 = data1[skip[0]:]
    data2 = data2[skip[1]:]
    if limit is not None:
        data1 = data1[:limit]
        data2 = data2[:limit]
    if data1 == data2:
        return None, IOResult()
    if silent:
        return None, IOResult(exit_code=1)
    common = min(len(data1), len(data2))
    if verbose:
        width = offset_width(sizes, limit)
        out_lines: list[str] = []
        for idx in range(common):
            if data1[idx] != data2[idx]:
                row = f"{idx + 1:>{width}} {data1[idx]:>3o}"
                if print_bytes:
                    row += f" {visible(data1[idx]):<4}"
                row += f" {data2[idx]:>3o}"
                if print_bytes:
                    row += f" {visible(data2[idx])}"
                out_lines.append(row)
        io = IOResult(exit_code=1)
        if len(data1) != len(data2):
            io.stderr = _eof_error(names, data1, data2, verbose)
        return format_records(out_lines), io
    for idx in range(common):
        if data1[idx] != data2[idx]:
            line = 1 + data1[:idx].count(_NEWLINE)
            # GNU counts in `byte` under -b and in `char` otherwise, on
            # the same offset -- the word tracks the flag, not a unit.
            unit = "byte" if print_bytes else "char"
            msg = (f"{names[0]} {names[1]}"
                   f" differ: {unit} {idx + 1}, line {line}")
            if print_bytes:
                msg += (f" is {data1[idx]:>3o} {visible(data1[idx])}"
                        f" {data2[idx]:>3o} {visible(data2[idx])}")
            return format_records([msg]), IOResult(exit_code=1)
    return None, IOResult(exit_code=1,
                          stderr=_eof_error(names, data1, data2, verbose))


def _eof_error(
    names: tuple[str, str],
    data1: bytes,
    data2: bytes,
    verbose: bool,
) -> bytes:
    """GNU's ``EOF on FILE`` diagnostic for a common-prefix difference.

    It is a diagnostic, not output: GNU writes it to stderr and still
    exits 1. A shorter file with no bytes to compare is ``which is
    empty``. Otherwise ``-l`` reports the byte only, and every other mode
    adds the line: ``line N`` when the file ends on a newline, ``in line
    N`` when it ends inside line N.

    Args:
        names (tuple[str, str]): the two operands as typed, in order.
        data1 (bytes): the first file's compared bytes.
        data2 (bytes): the second file's compared bytes.
        verbose (bool): whether ``-l`` is in effect.
    """
    shorter = names[0] if len(data1) < len(data2) else names[1]
    held = data1 if len(data1) < len(data2) else data2
    if not held:
        return f"cmp: EOF on {shorter} which is empty\n".encode()
    msg = f"cmp: EOF on {shorter} after byte {len(held)}"
    if not verbose:
        lines = held.count(_NEWLINE)
        msg += (f", line {lines}"
                if held[-1] == _NEWLINE else f", in line {lines + 1}")
    return (msg + "\n").encode()


__all__ = ["cmp_cmd"]


@dataclass(frozen=True, slots=True)
class CmpFlags:
    silent: bool = False
    verbose: bool = False
    limit: int | None = None
    print_bytes: bool = False
    skip: tuple[int, int] = (0, 0)


def parse_flags(flags: Mapping[str, FlagValue]) -> CmpFlags:
    fl = FlagView(flags, spec=SPECS["cmp"])
    n_raw = fl.as_str("n")
    i_raw = fl.as_str("i")
    return CmpFlags(
        silent=fl.as_bool("s"),
        verbose=fl.as_bool("args_l"),
        limit=parse_count(n_raw, "--bytes") if n_raw is not None else None,
        print_bytes=fl.as_bool("b"),
        skip=parse_skip(i_raw) if i_raw is not None else (0, 0),
    )


async def cmp_generic(
    paths: list[PathSpec],
    texts: list[str],
    opts: CommandOpts,
    read_bytes: Callable[..., Awaitable[bytes]],
) -> tuple[ByteSource | None, IOResult]:
    parsed = parse_flags(opts.flags)
    return await cmp_cmd(paths,
                         read_bytes=read_bytes,
                         stdin=opts.stdin,
                         silent=parsed.silent,
                         verbose=parsed.verbose,
                         limit=parsed.limit,
                         print_bytes=parsed.print_bytes,
                         skip=parsed.skip)

File: typescript/packages/core/src/commands/builtin/generic/cmp.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { specOf } from '../../spec/builtins.ts'
import { FlagView } from '../../spec/flag_view.ts'
import { IOResult, materialize, type ByteSource } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import type { CommandOpts } from '../../config.ts'
import { formatFsError, isFsError } from '../../../utils/errors.ts'
import { CMP_SIZE_UNITS, INTMAX, XSTRTOUMAX_PATTERN } from '../constants.ts'
import { STDIN_OPERAND } from '../utils/constants.ts'
import { formatRecords } from '../utils/output.ts'
import { isStdin, stdinStream } from '../utils/stream.ts'
import { parseBase0 } from '../utils/size_suffix.ts'
import { extraOperandError, missingOperandError, usageHint } from '../../spec/usage.ts'
import { CommandName } from '../../spec/types.ts'
import { UsageError } from '../../errors.ts'

const ENC = new TextEncoder()

const TRY_HELP = `\n${usageHint(CommandName.CMP)}`
const NEWLINE = 0x0a

function octal(n: number, width = 0): string {
  return n.toString(8).padStart(width)
}

/**
 * One GNU `cmp` byte count read the way xstrtoumax reads it.
 *
 * Base 0, so `010` is 8 and `0x400` is 1024; one leading `+` and
 * leading whitespace are allowed; the remainder is a size suffix from
 * cmp's own letter set. Every rejection -- unparsable digits, unknown
 * suffix, or a product past INTMAX -- is the same usage error naming
 * the long option, not a crash and not od's "too large".
 *
 * `shown` overrides the spelling named in the diagnostic: GNU prints
 * the operand from the position it was reading, so a bad `SKIP1` names
 * the whole `SKIP1:SKIP2` pair while a bad `SKIP2` names only itself.
 *
 * The accept/reject boundary is computed in BigInt and so is exact, but
 * the returned count is a double like od's: above 2**53 it is the
 * nearest representable value, not python's exact integer. The count is
 * only ever a slice bound, and no file reaches an exabyte, so the two
 * runtimes still compare the same bytes.
 */
export function parseCount(raw: string, option: string, shown?: string): number {
  const error = new UsageError(`cmp: invalid ${option} value '${shown ?? raw}'${TRY_HELP}`)
  const match = XSTRTOUMAX_PATTERN.exec(raw)
  const suffix = match?.[2] ?? ''
  const unit = suffix === '' ? 1 : CMP_SIZE_UNITS[suffix]
  if (match === null || unit === undefined) throw error
  const count = parseBase0(match[1] ?? '') * BigInt(unit)
  if (count > INTMAX) throw error
  return Number(count)
}

/**
 * The `-i` operand as one skip per file.
 *
 * GNU takes `SKIP` for both files or `SKIP1:SKIP2` for one each, so
 * `-i 0:3` compares all of the first file against the fourth byte
 * onward of the second. A colon is the only place the first count may
 * stop, which is why `1b:1` is rejected naming the whole pair while
 * `1:1b` is rejected naming just `1b`.
 */
export function parseSkip(raw: string): [number, number] {
  const cut = raw.indexOf(':')
  if (cut === -1) {
    const both = parseCount(raw, '--ignore-initial')
    return [both, both]
  }
  return [
    parseCount(raw.slice(0, cut), '--ignore-initial', raw),
    parseCount(raw.slice(cut + 1), '--ignore-initial'),
  ]
}

/**
 * One byte rendered the way GNU `cmp -b` renders it.
 *
 * The cat -v alphabet: a control byte becomes `^X` (so tab is `^I`,
 * unlike `cat -v` itself), DEL becomes `^?`, and a high byte becomes
 * `M-` followed by the same rules on its low seven bits.
 */
export function visible(byte: number): string {
  if (byte >= 128) return `M-${visible(byte - 128)}`
  if (byte === 127) return '^?'
  if (byte < 32) return `^${String.fromCharCode(byte + 64)}`
  return String.fromCharCode(byte)
}

interface CmpFlags {
  readonly silent: boolean
  readonly verbose: boolean
  readonly limit: number | null
  readonly printBytes: boolean
  readonly skip: readonly [number, number]
}

function parseFlags(fl: FlagView): CmpFlags {
  const nRaw = fl.asStr('n')
  const iRaw = fl.asStr('i')
  return {
    silent: fl.asBool('s'),
    verbose: fl.asBool('args_l'),
    limit: nRaw === undefined ? null : parseCount(nRaw, '--bytes'),
    printBytes: fl.asBool('b'),
    skip: iRaw === undefined ? [0, 0] : parseSkip(iRaw),
  }
}

function arraysEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.byteLength !== b.byteLength) return false
  for (let i = 0; i < a.byteLength; i++) if (a[i] !== b[i]) return false
  return true
}

/**
 * GNU's `EOF on FILE` diagnostic for a common-prefix difference.
 *
 * It is a diagnostic, not output: GNU writes it to stderr and still
 * exits 1. A shorter file with no bytes to compare is `which is empty`.
 * Otherwise `-l` reports the byte only, and every other mode adds the
 * line: `line N` when the file ends on a newline, `in line N` when it
 * ends inside line N.
 */
function eofError(
  names: readonly [string, string],
  data1: Uint8Array,
  data2: Uint8Array,
  verbose: boolean,
): Uint8Array {
  const firstShorter = data1.byteLength < data2.byteLength
  const shorter = firstShorter ? names[0] : names[1]
  const held = firstShorter ? data1 : data2
  if (held.byteLength === 0) return ENC.encode(`cmp: EOF on ${shorter} which is empty\n`)
  let msg = `cmp: EOF on ${shorter} after byte ${String(held.byteLength)}`
  if (!verbose) {
    let lines = 0
    for (const byte of held) if (byte === NEWLINE) lines += 1
    msg +=
      held[held.byteLength - 1] === NEWLINE
        ? `, line ${String(lines)}`
        : `, in line ${String(lines + 1)}`
  }
  return ENC.encode(`${msg}\n`)
}

/**
 * The width GNU `cmp -l` pads its offset column to.
 *
 * GNU sizes the column for the largest offset it could print: the `-n`
 * limit, cut to the bytes left in each regular file after its skip. A
 * stream has no size to cut by, so a line comparing two of them pads to
 * the width of the largest file offset.
 */
function offsetWidth(sizes: readonly number[], limit: number | null): number {
  let most = limit !== null ? BigInt(limit) : INTMAX
  for (const size of sizes) if (BigInt(size) < most) most = BigInt(size)
  return String(most > 0n ? most : 0n).length
}

export async function cmpGeneric(
  paths: PathSpec[],
  opts: CommandOpts,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
): Promise<[ByteSource | null, IOResult]> {
  const parsed = parseFlags(new FlagView(opts.flags, specOf('cmp')))
  if (paths.length > 2) throw extraOperandError(CommandName.CMP, paths[2]?.rawPath ?? '')
  const p0 = paths[0]
  if (p0 === undefined) throw missingOperandError(CommandName.CMP, null)
  // A lone FILE1 is compared with stdin, which GNU names `-`.
  const p1 = paths[1] ?? STDIN_OPERAND
  // Both name the one stdin: GNU sees the same file at the same offset
  // and answers equal without reading, whatever the skips.
  if (isStdin(p0) && isStdin(p1)) return [null, new IOResult()]
  const names = [p0.rawPath, p1.rawPath] as const
  const read = stdinStream(stream, opts.stdin)
  let data1: Uint8Array
  let data2: Uint8Array
  try {
    data1 = await materialize(read(p0))
    data2 = await materialize(read(p1))
  } catch (err) {
    if (!isFsError(err)) throw err
    // GNU cmp reserves exit 1 for "files differ"; trouble (a missing or
    // unreadable operand) is exit 2.
    return [null, new IOResult({ exitCode: 2, stderr: formatFsError('cmp', err, paths) })]
  }
  const sizes: number[] = []
  if (!isStdin(p0)) sizes.push(data1.byteLength - parsed.skip[0])
  if (!isStdin(p1)) sizes.push(data2.byteLength - parsed.skip[1])
  data1 = data1.slice(parsed.skip[0])
  data2 = data2.slice(parsed.skip[1])
  if (parsed.limit !== null) {
    data1 = data1.slice(0, parsed.limit)
    data2 = data2.slice(0, parsed.limit)
  }
  if (arraysEqual(data1, data2)) return [null, new IOResult()]
  if (parsed.silent) return [null, new IOResult({ exitCode: 1 })]
  const common = Math.min(data1.byteLength, data2.byteLength)
  if (parsed.verbose) {
    const width = offsetWidth(sizes, parsed.limit)
    const outLines: string[] = []
    for (let idx = 0; idx < common; idx++) {
      const a = data1[idx] ?? 0
      const b = data2[idx] ?? 0
      if (a === b) continue
      let row = `${String(idx + 1).padStart(width)} ${octal(a, 3)}`
      if (parsed.printBytes) row += ` ${visible(a).padEnd(4)}`
      row += ` ${octal(b, 3)}`
      if (parsed.printBytes) row += ` ${visible(b)}`
      outLines.push(row)
    }
    const io =
      data1.byteLength === data2.byteLength
        ? new IOResult({ exitCode: 1 })
        : new IOResult({ exitCode: 1, stderr: eofError(names, data1, data2, true) })
    return [formatRecords(outLines), io]
  }
  for (let idx = 0; idx < common; idx++) {
    const a = data1[idx] ?? 0
    const b = data2[idx] ?? 0
    if (a === b) continue
    let line = 1
    for (let k = 0; k < idx; k++) if (data1[k] === NEWLINE) line += 1
    // GNU counts in `byte` under -b and in `char` otherwise, on the
    // same offset -- the word tracks the flag, not a unit.
    const unit = parsed.printBytes ? 'byte' : 'char'
    let msg = `${names[0]} ${names[1]} differ: ${unit} ${String(idx + 1)}, line ${String(line)}`
    if (parsed.printBytes) {
      msg += ` is ${octal(a, 3)} ${visible(a)} ${octal(b, 3)} ${visible(b)}`
    }
    return [formatRecords([msg]), new IOResult({ exitCode: 1 })]
  }
  return [
    null,
    new IOResult({ exitCode: 1, stderr: eofError(names, data1, data2, parsed.verbose) }),
  ]
}
